recipes/LibriSpeech/ASR/CTC/train_with_wav2vec_k2.py

```diff
@@ -0,0 +1,656 @@
+        if hasattr(self.modules, "extractor"):  # SpeechBrain pretrained model
+            latents = self.modules.extractor(wavs)
+            feats = self.modules.encoder_wrapper(latents, wav_lens=wav_lens)[
+                "embeddings"
+            ]
+        else:  # HuggingFace pretrained model
+            feats = self.modules.wav2vec2(wavs, wav_lens)
+
+        x = self.modules.enc(feats)
+
+        # Compute outputs
+        logits = self.modules.ctc_lin(x)
+
+        # Upsample the inputs if they have been highly downsampled
+        if hasattr(self.hparams, "upsampling") and self.hparams.upsampling:
+            logits = logits.view(
+                logits.shape[0], -1, self.hparams.output_neurons
+            )
+
+        p_ctc = self.hparams.log_softmax(logits)
+        return p_ctc, wav_lens
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the loss (CTC+NLL) given predictions and targets."""
+
+        p_ctc, wav_lens = predictions
+
+        ids = batch.id
+
+        # Sort batch to be descending by length of wav files, which is demanded by k2
+        if self.hparams.sorting == "ascending":
+            p_ctc = torch.flip(p_ctc, (0,))
+            wav_lens = torch.flip(wav_lens, (0,))
+            texts = [batch.wrd[i] for i in reversed(range(len(batch.wrd)))]
+        elif self.hparams.sorting == "descending":
+            texts = batch.wrd
+        else:
+            raise NotImplementedError(
+                "Only ascending or descending sorting is "
+                "implemented, but got {}".format(self.hparams.sorting)
+            )
+
+        is_training = stage == sb.Stage.TRAIN
+        loss = self.hparams.ctc_cost(
+            log_probs=p_ctc,
+            input_lens=wav_lens,
+            graph_compiler=self.graph_compiler,
+            texts=texts,
+            is_training=is_training,
+        )
+
+        if stage == sb.Stage.VALID:
+            # Decode token terms to words
+            predicted_texts = self.graph_compiler.decode(
+                p_ctc,
+                wav_lens,
+                ac_scale=self.hparams.ac_scale,
+                decoding_method="1best",
+            )  # list of strings
+            predicted_words = [wrd.split(" ") for wrd in predicted_texts]
+            target_words = [wrd.split(" ") for wrd in texts]
+            self.wer_metric.append(ids, predicted_words, target_words)
+            self.cer_metric.append(ids, predicted_words, target_words)
+        if stage == sb.Stage.TEST:  # Language model decoding only used for test
+            decoding_method = self.hparams.decoding_method
+            # If the decoding method is 1best then the metric stats will be
+            # saved in a single file, otherwise, a new directory will be created
+            # for each lm_scale used in whole lattice rescoring.
+            decode_output: Union[dict, List[str]] = self.graph_compiler.decode(
```

Comment:
This notation is uncommon in SB. Is the typing necessary here? Otherwise, I would advise removing it.

Reply:
I think it can help users understand the code without having to read all of the comments (i.e. in this case, understand that there are only two types this variable could have).

I would be happy to remove it if you think it looks bad, though. Should I remove every kind of typing or only the `Union`-based ones?